### PR TITLE
Add dynamic imports

### PR DESCRIPTION
**Pull request by LeonMrBonnie**
The callback for waiting until all resources are loaded does not work yet, but the implementation for dynamic import works.

### CODE REVIEW

**Comment by FabianTerhorst** [comment]
I don't think its a good idea to use the js script runtime for detecting when all resources are loaded.
Better request a api for that in cpp sdk. Otherwise dynamic imports for different resource types like asset won't work.

**Comment by FabianTerhorst** [comment]
Its also not clean to use the connection complete event for that i think.

**Comment by LeonMrBonnie** [comment]
> 
> 
> I don't think its a good idea to use the js script runtime for detecting when all resources are loaded.
> Better request a api for that in cpp sdk. Otherwise dynamic imports for different resource types like asset won't work.

Yes, I know. But currently there is no other way except than using the connectionComplete event.
That's why this is a draft, if a callback for that could be implemented in C++ SDK then it would be even better.

File: src/CV8ScriptRuntime.h [hunk 1]
```diff
@@ -127,4 +127,37 @@ class CV8ScriptRuntime : public alt::IScriptRuntime
...
+		}
+
+		v8::Local<v8::ScriptOrModule> _referrer;
+		v8::Local<v8::String> _specifier;
+		ResourcesLoadedCallback _callback;
+		const void* _promise;
+
+		void call()
+		{
+			if (&_promise == nullptr) return;
```
Comment (Unnvaldr) [inline src/CV8ScriptRuntime.h]: What did you intend to do here with getting address of pointer `_promise` and comparing it against null pointer keyword?

Comment (LeonMrBonnie) [reply src/CV8ScriptRuntime.h]: Leftover code from debugging, removed it

**Comment by Unnvaldr** [comment]
> I don't think its a good idea to use the js script runtime for detecting when all resources are loaded.
> Better request a api for that in cpp sdk. Otherwise dynamic imports for different resource types like asset won't work.

While agree with that, I think that it can land anyway in dev branch for bigger audience to test it.

File: src/CV8Resource.cpp [hunk 2]
```diff
@@ -235,6 +235,10 @@ bool CV8ResourceImpl::OnEvent(const alt::CEvent *e)
 	case alt::CEvent::Type::CONNECTION_COMPLETE:
 	{
 		handlers = GetLocalHandlers("connectionComplete");
+		CV8ScriptRuntime* runtime = CV8ScriptRuntime::instance;
+		runtime->_allResourcesLoaded = true;
+		for (CV8ScriptRuntime::ResourcesLoadedResult result : runtime->onAllResourcesLoadedCallbacks) result.call();
+		runtime->onAllResourcesLoadedCallbacks.empty();
```
Comment (7Hazard) [inline src/CV8Resource.cpp]: This operation .empty() has no effect here, I suppose you meant .clear()

File: src/CV8ScriptRuntime.cpp [hunk 3]
```diff
@@ -58,10 +58,64 @@ CV8ScriptRuntime::CV8ScriptRuntime()
 		}
 	});
 
-	/*isolate->SetHostImportModuleDynamicallyCallback([](v8::Local<v8::Context> context, v8::Local<v8::ScriptOrModule> referrer, v8::Local<v8::String> specifier) {
-
-		return v8::MaybeLocal<v8::Promise>();
-	});*/
+	static std::list<v8::UniquePersistent<v8::Promise::Resolver>> promises;
```
Comment (7Hazard) [inline src/CV8ScriptRuntime.cpp]: Name it dynamicImportsPromises

File: src/CV8ScriptRuntime.h [hunk 4]
```diff
@@ -127,4 +127,36 @@ class CV8ScriptRuntime : public alt::IScriptRuntime
 			return result;
 		}
 	};
+
+	using ResourcesLoadedCallback = void (*)(v8::Local<v8::ScriptOrModule> referrer, v8::Local<v8::String> specifier, const void* promise);
+
+	class ResourcesLoadedResult
```
Comment (7Hazard) [inline src/CV8ScriptRuntime.h]: Could be named DynamicImport cause this class (at least for now) is quite specific to dynamic imports

**Review by 7Hazard — CHANGES_REQUESTED** [review]
1.
Dynamic imports are not properly resolved after exactly all resources has loaded:
```
[06:31:13] Loading resource example
[06:31:13][Debug] Created Thread with ID: 21
[06:31:13] [V8] Starting script client/src/main.js
[06:31:14] DID STUFF
[06:31:14][Debug] Sent event test 11
[06:31:14] DYNAMIC IMPORT
[06:31:14] DID STUFF
[06:31:14] [V8] Started script client/src/main.js
[06:31:14] Loaded resource example
```
As visible here, 'DID STUFF', which is a log from the to-be imported module is called before the resource itself has finished loading the resource 'example'. Perhaps this was the intent for modules inside the same resource.

**Pull request merged**